### PR TITLE
Use idempotent JSON patch updates to avoid "object has been modified"

### PR DESCRIPTION
**Pull request by andyhd**
Using Kubernetes API `update` to replace a deployment or ingress can cause collateral damage to metadata, which Kubernetes complains about as "object has been modified".

screenshot 2019-01-09 at 15 14 31

The fix is to use `patch` updates to change only the values necessary. These patches are idempotent, so remove the need to fetch and re-fetch the resources.

Also shown in the error image above, error messages for a host other than the one the user is trying to unidle may be shown. The fix is to send SSE messages only to groups of clients subscribing to events for a specific host. Eg: a user requests to unidle "foo.example.com" and is added to a group of clients who will receive messages about "foo.example.com", but not "bar.example.com".

### CODE REVIEW

**Comment by andyhd** [comment]
That's a good point that the race condition is still possible (although with a much reduced window).

I think the problem is actually the way I am updating - replacing the whole deployment or ingress instead of patching them. The "object has been modified" error message is happening because I'm overwriting metadata which is irrelevant to the operation I want to perform.

I think you had it right originally - I'll change the update code to patch instead.

**Comment by xoen** [comment]
I guess patch is better because it only change a tiny portion of the resource so your chances of getting a version conflict from k8s should be less. But I'm not a k8s expert by all means.

File: Makefile [hunk 1]
```diff
@@ -34,6 +34,11 @@ test: dependencies
 	@echo " > Testing..."
 	@${GO} test -v
 
+## race-check: Check for race conditions
```
Comment (xoen) [inline Makefile]: ❤️

File: unidler.go [hunk 2]
```diff
@@ -13,50 +13,56 @@ type (
 		host             string
```
Comment (xoen) [inline unidler.go]: Not that it changes anything for the complier or CPU but just a *minor* style comment: as you're only defining only a type here, use this form is overkill, so you may want to write `type Unidler struct {` in a single line.

(but as I said, it's minor so feel free to ignore!)

File: k8s.go [hunk 3]
```diff
@@ -113,7 +111,7 @@ func (k KubernetesAPI) Deployment(ing *v1beta1.Ingress) (*v1.Deployment, error)
 	}
 
 	dep := &deployments.Items[0]
-	k.log(fmt.Sprintf("Deployment found '%s' (ns: '%s')\n", dep.Name, dep.Namespace))
+	k.log.Printf("Deployment found '%s' (ns: '%s')\n", dep.Name, dep.Namespace)
```
Comment (xoen) [inline k8s.go]: ❤️

File: main.go [hunk 4]
```diff
@@ -37,9 +41,14 @@ func main() {
 	}
 
 	// parse HTML template
-	tmpl, err := template.ParseFiles("templates/index.html")
+	tmpl, err := template.New("").ParseFiles(
```
Comment (xoen) [inline main.go]: I really like that you split this in smaller parts ❤️ 

I have a couple of observations and I propose to rename these slightly (if you want/agree of course!):
- `base.html` => `layout.html` (defining `layout`)
- `index.html` => (maybe) rename to `content.html` which defines `content` (or maybe call them `body.html`/`body`)
- `index.js` => (maybe) rename to `javascript.js` which defines `javascript`

It took me few more seconds to process/map which file was defining which template, etc...

Basically I would, IMHO of course, at least:
1. rename `base` to `layout` (both file and template name).
2. rename either the file *or* the template name of `index.html`/`content` so that they match.

Comment (xoen) [reply main.go]: (also having matching file and template name helps when jumping between files, or at least in VSCode/Sublime)

File: main.go [hunk 5]
```diff
@@ -72,7 +81,7 @@ func healthCheck(w http.ResponseWriter, req *http.Request) {
 
 func (u *Unidler) unidle(w http.ResponseWriter, req *http.Request) {
```
Comment (xoen) [inline main.go]: I wonder if it would be predictable to move this method into the `unidler.go` file. I know it's still the same package (`main`) but I keep looking for it in the `unidler.go` (where the other `Unidler`'s methods are defined) before I realise is actually defined in `main.go`.

File: sse.go [hunk 6]
```diff
@@ -57,27 +96,95 @@ func (s *SSEBroker) ServeHTTP(w http.ResponseWriter, req *http.Request) {
...
+	group := s.getOrCreateGroup(client.group)
+	group.add(client)
+	s.log.Printf("Added client to group %s", group.name)
+
+	if group.lastMessage != nil {
+		client.send <- group.lastMessage
+	}
+}
+
+func (s *SSEBroker) getOrCreateGroup(name string) *Group {
```
Comment (xoen) [inline sse.go]: I still have a bad feeling about this map. Bear in mind that potentially two requests for two different hosts/groups may arrive "concurrently" and the assignment may be forced to allocate more memory/grow the map for the additional element.

If two different go routines create two different maps you may have a race condition. If I understand correctly one of these two "new" maps/memory would be lost which means one of the two groups may not be present in the final map attached to the `SSEBroker` instance (struct field `groups`).

...but maybe I'm talking rubbish 🤷‍♂️ 

I got a bad feeling about this

**Comment by andyhd** [comment]
I've left some TODOs in because we can simplify further using a `host` label on the app deployments and ingresses, but they are not in place yet. Also, I'm testing modifying an app service to point to the unidler deployment, rather than modifying ingresses, which may also allow removing a bunch of code.

**Comment by andyhd** [comment]
Good point, though - I will create some tickets in Trello

**Comment by xoen** [comment]
@andyhd I was joking really, but of course if you want to keep trace of this feel free.
